### PR TITLE
Up and down arrow in the console moves through command history

### PR DESCRIPTION
**Pull request by jcmorrow**
### Motivation

As mentioned in #3842, a common expectation of consoles is that you can access previous commands by hitting the up arrow. This will add support within a session for traversing history with the arrows.

### Related Changes

While in the messages reducer I tried to hack away at some of the dead code around things like message groups and traces.

### Future Work

I've had this part working for a while, but have spent a couple of hours working on getting history to persist across refresh and page navigations. That is considerably more complicated, and I'm still working on it, but I wanted to get this shipped as fast as possible, since it's useful even in its current form.

### Screenshot

https://user-images.githubusercontent.com/5903784/136268630-b894366a-31e5-403c-be6f-e5310cd16bad.mp4

### CODE REVIEW

File: src/devtools/client/webconsole/selectors/messages.js [hunk 1]
```diff
@@ -48,6 +48,10 @@ export const getMessages = createSelector(
   (messagesById, visibleMessages) => visibleMessages.map(id => messagesById.get(id))
 );
 
+export const getCommandMessages = createSelector(getMessages, messages =>
+  messages.filter(message => message.type === "command")
+);
```
Comment (jasonLaster) [inline src/devtools/client/webconsole/selectors/messages.js:53]: that's really smart

File: src/devtools/client/webconsole/components/Input/JSTerm.js [hunk 2]
```diff
@@ -59,20 +65,38 @@ class JSTerm extends React.Component {
...
+      historyCursorDown: this.historyCursorDown,
+    });
     this.editor = await this.editorWaiter;
     this.editor.appendToLocalElement(this.node);
 
     const recordingId = getRecordingId();
     const recording = await getRecording(recordingId);
 
-    this.setState({ canEval: recording.userRole !== "team-user" });
+    this.setState({ canEval: recording.userRole !== "team-user", historyIndex: 0 });
```
Comment (jasonLaster) [inline src/devtools/client/webconsole/components/Input/JSTerm.js:79]: 👍

Comment (jaril) [inline src/devtools/client/webconsole/components/Input/JSTerm.js:79]: Small nit: any reason why we couldn't just set the `historyIndex` in the constructor?

File: src/devtools/client/webconsole/components/Input/JSTerm.js [hunk 3]
```diff
@@ -59,20 +65,38 @@ class JSTerm extends React.Component {
...
+  historyCursorUp = () => this.moveHistoryCursor(1);
+  historyCursorDown = () => this.moveHistoryCursor(-1);
+
+  moveHistoryCursor(difference) {
+    const messages = uniq(this.props.commandHistory.map(m => m.messageText).reverse());
+
+    if (messages.length === 0) {
+      return;
+    }
+    const newIndex = clamp(this.state.historyIndex + difference, 0, messages.length);
```
[the remark this answers is not in the document]
Comment (jcmorrow) [reply src/devtools/client/webconsole/components/Input/JSTerm.js:95]: I always want it to be `clamp(min, candidate, max)`, so I always get the ordering wrong 😭

[the remark this answers is not in the document]
Comment (jaril) [reply src/devtools/client/webconsole/components/Input/JSTerm.js:95]: Shouldn't this be `messages.length - 1` here?

[the remark this answers is not in the document]
Comment (jaril) [reply src/devtools/client/webconsole/components/Input/JSTerm.js:95]: Ah wait I just saw line 96. Sick 👍

**Review by jaril — APPROVED** [review]
Played around with it and it seems to work well. Ship it!

**Pull request merged**